File: .ci/examples/py-examples/exclamation.py
from pulsar import Function

# The classic ExclamationFunction that appends an exclamation at the end
# of the input
class ExclamationFunction(Function):
  def __init__(self):
    pass

  def process(self, input, context):
    logger = context.get_logger()
    msg_id = context.get_message_id()
    logger.info("processing message %s in function %s/%s/%s" % (msg_id, context.get_function_tenant(), context.get_function_namespace(), context.get_function_name()))
    for word in input.split(" "):
      context.incr_counter(word, 1)
      count = context.get_counter(word)
      logger.info("got word: %s for %d times" % (word, count.value))
    context.publish("persistent://public/default/test-py-package-serde-extra", input + '!')
    return input + '!'

File: .ci/examples/node-examples/exclamation.js
async function process(message, context) {
    const logger = context.getLogger();
    let msg_id = await context.getMessageId()
    logger.info(`processing message ${msg_id} in function ${context.getFunctionTenant()}/${context.getFunctionNamespace()}/${context.getFunctionName()}`)
    for (let word of message.split(" ")) {
        await context.incrementCounter(word, 1)
        count = await context.getCounter(word)
        logger.info(`got word: ${word} for ${count['value']} times`)
    }
    await context.publish("persistent://public/default/test-node-package-serde-extra", message.concat("!!!"))
    return message.concat("!");
}
